src/btree/bt_page.c

```diff
@@ -120,240 +120,427 @@ __page_build_ref(WT_SESSION_IMPL *session, WT_REF *parent_ref, WT_CELL_UNPACK_AD
     int cmp;
+    bool row_leaf_page = false, internal_page = false;
+
+    WT_UNUSED(incr);
+    WT_UNUSED(p);
 
-    key_a.data = a->key.data;
-    key_a.size = a->key.size;
-    key_b.data = b->key.data;
-    key_b.size = b->key.size;
+    unpacked_deltas = NULL;
+    delta_size_each = NULL;
+    refs = NULL;
+    final_entries = 0;
+    page = ref->page;
 
-    if ((ret = __wt_compare(session, S2BT(session)->collator, &key_a, &key_b, &cmp)) != 0)
-        WT_IGNORE_RET(__wt_panic(session, ret, "failed to compare keys"));
+    /* Step 1: Unpack deltas. */
+    WT_RET(__page_unpack_deltas_old(
+      session, page, deltas, delta_size, &unpacked_deltas, &delta_size_each));
 
-    return (cmp);
-}
+    /* Step 2: Unpack base entries. */
+    k = 0;
+    base_entries = page->dsk->u.entries;
+    WT_ERR(__wt_calloc_def(session, base_entries, &base));
+    WT_CELL_FOREACH_ADDR (session, page->dsk, base[k]) {
+        k++;
+    }
+    WT_CELL_FOREACH_END;
 
-/*
- * __page_merge_internal_deltas --
- *     Perform a k-way merge sort on the delta arrays to create a single consolidated delta array.
- */
-static int
-__page_merge_internal_deltas(WT_SESSION_IMPL *session, WT_CELL_UNPACK_DELTA_INT **unpacked_deltas,
-  uint32_t start, uint32_t end, size_t *sizes, WT_CELL_UNPACK_DELTA_INT ***deltas_merged,
-  size_t *size)
-{
-    WT_CELL_UNPACK_DELTA_INT **left, **merged, **right;
-    size_t left_size, right_size;
-    uint32_t i, mid;
-
-    WT_ASSERT(session, start <= end);
-    left = right = merged = NULL;
-
-    if (start == end) {
-        *size = sizes[start];
-        WT_RET(__wt_calloc_def(session, *size, &merged));
-        for (i = 0; i < (uint32_t)sizes[start]; ++i)
-            merged[i] = &unpacked_deltas[start][i];
-        *deltas_merged = merged;
-        return (0);
+    /* Step 3: Estimate max refs. */
+    estimated_entries = (base_entries / 2) + 1;
+    for (d = 0; d < delta_size; ++d)
+        estimated_entries += delta_size_each[d];
+    WT_ERR(__wt_calloc_def(session, estimated_entries, &refs));
+
+    /* Step 4: Initialize disk image header. */
+    WT_ERR(__wt_buf_init(session, disk_image, page->dsk->mem_size));
+    WT_ERR(__wt_buf_extend(session, disk_image, sizeof(WT_PAGE_HEADER)));
+    hdr = (WT_PAGE_HEADER *)disk_image->data;
+    memset(hdr, 0, sizeof(WT_PAGE_HEADER));
+    hdr->type = page->type;
+    p = (uint8_t *)hdr + sizeof(WT_PAGE_HEADER);
+
+    /* Step 5: Determine page type. */
+    switch (ref->page->type) {
+    case WT_PAGE_ROW_LEAF:
+        row_leaf_page = true;
+        break;
+    case WT_PAGE_ROW_INT:
+        internal_page = true;
+        break;
+    default:
+        return (__wt_illegal_value(session, ref->page->type));
     }
 
-    mid = (start + end) / 2;
+    /* Step 6: Merge loop using common helper. */
+    WT_RET(__wt_calloc_def(session, delta_size, &delta_idx));
+    i = 0;
+    while (true) {
+        WT_CELL_UNPACK_DELTA_INT *min_delta = NULL;
+        uint32_t min_d = UINT32_MAX;
```

Comment:
We also need to merge with WT-15269 change.

Reply:
Fixed in c714ce0f95d318db5687b8fe0308bdbeaa7c4d1a